### ISSUE TITLE
Ability to add components deployed in other namespaces

### ISSUE THREAD

**Issue by durandom** [opendatahub-io/odh-dashboard#26]
**Is your feature request related to a problem? Please describe.**
the dashboard seems to be limited to the namespace where it's deployed.

**Describe the solution you'd like**
All components deployed by the operator should be included, independent of the namespace they're deployed to

https://github.com/operate-first/odh/issues/12
https://github.com/operate-first/odh/issues/13

**Comment by cfchase** [comment]
I may be incorrect in the way I'm interpreting ODH installations, so I'll have to confer with @vpavlin and @nakfour.

In general, I think operators deploy instances and that instance would be the owner of and resources of their CRD would be related to that instance.  I don't know if the ODH operator establishes that relationship, but usually that relationship exists.

The dashboard currently assumes it was installed as a component along with the instance of ODH it was installed with.  The ODH operator then installs the components in that single namespace.  Another installation in another namespace, in my current understanding, is a separate installation, not related to this dashboard or the instance it's a part of.

Your request assumes there's a multiple installations (of single components) across several namespaces and should all show up on the same dashboard as part of a single ODH installation on a cluster.  Maybe @vpavlin and @nakfour can establish whether this is a safe assumption.

**Comment by vpavlin** [comment]
IIUC this is what @durandom's team is trying to achieve - by using multiple KFDefs install various parts of ODH into multiple namespaces to separate the components from each other.

I think it might be problematic if we allow this "freely" - in case 2 users decide to deploy 2 instances of JupyterHub to the same cluster (which should work fine) - would both dashboard show both instances? That could bring some security concerns

If we want to allow this those KFDefs that should be bound together have to have something in common - a label or an annotation at least, otherwise it would get very confusing and potentially dangerous.

**Comment by cfchase** [comment]
Perhaps this would be better off in a fork in the operate-first org, for now.

I haven't encountered kubernetes resources that have components that spans separate namespaces.  If what you're telling me is a goal, then perhaps it would be better for operate-first to have it's own map or custom resource that points to each of the sub-components  as owner.  Then the dashboard would query the operate-first resource for the kfdefs and routes from its subcomponents.

This application where it treats all resources across the cluster as part of a single entity seems like it would be hard to integrate with the general community offering right now and might cause issues.

**Comment by durandom** [comment]
@vpavlin we had a lengthy discussion on single-namespace deploy vs multi-namespace deploy at https://github.com/operate-first/odh/issues/13 and in google chat (which I cannot link, unfortunately).

Almost everybody said, in a production grade environment you'd likely won't do a single-namespace deploy.

May I suggest to provide guidance via an ADR (https://github.com/opendatahub-io/odh-manifests/pull/224) from ODH?
Then re-visit this issue/PR

**Comment by vpavlin** [comment]
I kind of like @cfchase's idea - at least as an interim solution - basically having a top-level CR (and probably a ismple operator to handle it) which would create and manage multiple KFDefs based on that...on the other hand it adds another layer of indirection and complexity...

Can we discuss this again at some ODH (community) meeting? I think it is valuable discussion - how do we want to approach multi-namespace deployment - since that is something Kubeflow does not do, so if we need some changes to support it, they might need to go there first.

**Comment by vpavlin** [comment]
IOW I think if we want to create a concept of multi-namespaced deployment for ODH which is represented by code and UI, we need to design it properly. Multiple KFDefs approach was more a workaround to unblock operate-first initiative. Hope that makes sense

**Comment by tumido** [comment]
Can you please loop me in for that discussion @vpavlin ?

I don't think that's correct assumption @cfchase, you definitely can have a controller which is expected to touch different namespaces. This is a common design pattern, actually the ODH operator leverages this itself as the actual operator is deployed in one namespace and manages resources in another namespaces.

There are 2 things happening here though. A `kfdef` resource is [a namespaced resource](https://github.com/kubeflow/kfctl/blob/cee16ff8a0eccee9333cbe72eaf1cc7b422aa651/deploy/crds/kfdef.apps.kubeflow.org_kfdefs_crd.yaml#L12), therefore user can deploy as many instances as they want in their cluster. They can have multiple `kfdef` resources in a single namespace, they also can have another `kfdef`s in other namespaces. The kubeflow operator handles this transparently and takes care of all the resources of that kind it finds.

ODH Dahsboard works on a "visibility range" which is defined to it via its service account. This SA has a role assigned allowing it to list the `kfdef`s in a specific namespace. By extending this role assignment to another namespaces, you can make the same dashboard to see `kfdef` resources in these other namespaces as well. See my implementation in #27. This implementaion doesn't handle the case when you have the very same component deployed multiple times, but it's solvable by reversing the `availableComponent` traversal.

**Comment by cfchase** [comment]
> Almost everybody said, in a production grade environment you'd likely won't do a single-namespace deploy.
> 
To clarify, multi-namespace vs single isn't really my issue.  My issue is treating 8 separate instances of a custom resource as a single entity seems bad.

**Comment by tumido** [comment]
Btw, I've just checked how upstream is doing this in the central dashboard of kubeflow. Instead of looking up the `kfdef`'s the operator builds a configmap for each namespace ([this is the default state](https://github.com/kubeflow/kubeflow/blob/master/components/centraldashboard/config/centraldashboard-links-config.yaml) where the kubeflow is deployed. And this configmap is then consumed from the cluster-wide "central dashboard" deployment, where user select the namespace he's talking to and that's it. Would something like that work for ODH?

**Comment by cfchase** [comment]
> Btw, I've just checked how upstream is doing this in the central dashboard of kubeflow. Instead of looking up the `kfdef`'s the operator builds a configmap for each namespace ([this is the default state](https://github.com/kubeflow/kubeflow/blob/master/components/centraldashboard/config/centraldashboard-links-config.yaml)) where the kubeflow is deployed. And this configmap is then consumed from the cluster-wide "central dashboard" deployment, where user select the namespace he's talking to and that's it. Would something like that work for ODH?

I think that was similar to what I was suggesting with the Custom Resource, but using a config map.

**Comment by HumairAK** [comment]
> To clarify, multi-namespace vs single isn't really my issue. My issue is treating 8 separate instances of a custom resource as a single entity seems bad. 

Isn't this precisely what the [notebook-controller](https://github.com/kubeflow/kubeflow/tree/master/components/notebook-controller#spec) does in kubeflow with the `notebook` resource? The prometheus operator also does the same thing with `service-monitors` 

Anyways besides concerns with the dashboard, there are other issues with using multi-namespace installations odh that we're currently encountering, more specifically with monitoring. The gist of it is that ocp now recommends to not use the `targetNamespaces` to extend operator access via `OperatorGroup` (see [here](https://docs.openshift.com/container-platform/4.6/operators/understanding/olm/olm-understanding-operatorgroups.html#olm-operatorgroups-target-namespace_olm-understanding-operatorgroups) ) and that was afaik the only way we could give prometheus access to other namespace to access the service monitors (so it's either one namespace or all namespaces, but all namespaces was not working for me). I'll make a more detailed issue in odh-manifests detailing it more thoroughly so as not to derail this discussion.

**Comment by cfchase** [comment]
> > To clarify, multi-namespace vs single isn't really my issue. My issue is treating 8 separate instances of a custom resource as a single entity seems bad.
> 
> Isn't this precisely what the [notebook-controller](https://github.com/kubeflow/kubeflow/tree/master/components/notebook-controller#spec) does in kubeflow with the `notebook` resource? The prometheus operator also does the same thing with `service-monitors`
The notebook controller looks like it establishes a relationship in that spec... right there.  There's no relationship between these kfdef resources right now.

> Anyways besides concerns with the dashboard, there are other issues with using multi-namespace installations odh that we're currently encountering, more specifically with monitoring. The gist of it is that ocp now recommends to not use the `targetNamespaces` to extend operator access via `OperatorGroup` (see [here](https://docs.openshift.com/container-platform/4.6/operators/understanding/olm/olm-understanding-operatorgroups.html#olm-operatorgroups-target-namespace_olm-understanding-operatorgroups) ) and that was afaik the only way we could give prometheus access to other namespace to access the service monitors (so it's either one namespace or all namespaces, but all namespaces was not working for me). I'll make a more detailed issue in odh-manifests detailing it more thoroughly so as not to derail this discussion.

I think that's my point, the dashboard is just trying to follow what the operator does, and currently it installs all components in a single namespace per instance.  If you change that to relate separate resources in separate namespaces, the dashboard will follow.

**Comment by HumairAK** [comment]
Ah all right that makes more sense, this is basically I think what argocd does with `AppProject` and `Application` custom resources, where `Applications` _belong_ to an `AppProject` custom resource.

**Comment by cfchase** [comment]
Having odh-configuration resource that lists namespaces, routes, components, etc would be really useful for https://github.com/opendatahub-io/odh-dashboard/issues/21. I was considering adding my own configmap, but was thinking it would be better to be created by odh itself so it would be a single source of truth for the odh instance and the dashboard.

Having a kfdef per component is interesting as it allows the operator to modify an odh installation and remove a single component by deleting that kfdef.  Currently, I think that's problematic, right @LaVLaS ?

**Comment by tumido** [comment]
Based on our proposal in https://github.com/opendatahub-io/odh-manifests/issues/236 we'd like to support multiple namespaces via namespace parameter in the `kfdef`. Translated: ODH upstream would like to keep a single `kfdef` resource, while allowing the operator to deploy into mutliple namespace. So the logic for `odh-dashboard` would change a bit:

There's still expected to be a single `kfdef` resource and `odh-dashboard` has to have access to it (either it's in the same namespace or the namespace where `kfdef` can be found has to be forwarded to the dashboard/granted read access).

Dashboard has to parse the namespaces out of the `kfdef` resource and look for the routes in those namespaces (again, it has to have access to these namespace to list routes).

**Comment by andrewballantyne** [comment]
This issue is 2 years old. Closing due to being stale. Looks like we might have supported some of this.

If this is still a desire and we have not handled it to your desired state, please reopen and we can start the conversation again.
